### PR TITLE
Week 5

### PR DESCRIPTION
**Pull request by savithaR24**
Main Screen displays all countries with their capitals as a List which are sourced from REST Countries API. Loading icon is shown when communicating with the network to get data. If no Wi-fi, then an error is shown and RETRY option provided. On selecting a country, the user is taken to the detail screen which shows its capital, population, area and its flag. On pressing the back button, the user is taken back to the list. Works for Portrait/Landscape and Dark/Light mode combos.

### CODE REVIEW

File: app/src/main/java/com/kodeco/android/countryinfo/api/CountryAPIService.kt [hunk 1]
```diff
@@ -0,0 +1,10 @@
+package com.kodeco.android.countryinfo.api
+
+import com.kodeco.android.countryinfo.models.Country
+import retrofit2.Response
+import retrofit2.http.GET
+
+interface CountryAPIService {
+    @GET("all")
+    suspend fun getAllCountries(): Response<List<Country>>
```
Comment (zlippard) [inline app/src/main/java/com/kodeco/android/countryinfo/api/CountryAPIService.kt]: nitpick but you could remove the `Response` wrapper here and just return `List<Country>` type.

File: app/src/main/java/com/kodeco/android/countryinfo/models/Country.kt [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+data class Country(
+    val name: CountryName,
+    val capital: List<String>?,
+    val population: Long,
+    val area: Double,
+    val flags: CountryFlags,
+) {
+    val commonName = name.common
+    val flagUrl = flags.png
+    val nonNullCapital = capital?.firstOrNull() ?: "N/A"
```
Comment (zlippard) [inline app/src/main/java/com/kodeco/android/countryinfo/models/Country.kt:15]: Great job using a default value here!

File: app/src/main/java/com/kodeco/android/countryinfo/ui/components/CountryErrorScreen.kt [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    Column(
+        modifier = Modifier
+            .fillMaxSize()
+            .padding(20.dp),
+        horizontalAlignment = Alignment.CenterHorizontally,
+        verticalArrangement = Arrangement.Center
+    ) {
+        var customError = error.message
+        if (customError != null && customError.contains("Unable to resolve host")) {
+            customError = "No Network Connection. Please Check Connection and Try Again!"
```
Comment (zlippard) [inline app/src/main/java/com/kodeco/android/countryinfo/ui/components/CountryErrorScreen.kt:31]: While this works, be careful using string contains checks to associate certain error types. If the SDK ever changes their text for host resolution then this check won't work.

File: app/src/main/java/com/kodeco/android/countryinfo/ui/components/CountryInfoScreen.kt [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+                    val countriesResponse = service.getAllCountries()
+
+                    if (countriesResponse.isSuccessful) {
+                        CountryInfoState.Success(countriesResponse.body()!!)
+                    } else {
+                        CountryInfoState.Error(Throwable("Request Failed - ${countriesResponse.message()}"))
+                    }
+                } catch (exception: Exception) {
+                    CountryInfoState.Error(exception)
+                }
```
Comment (zlippard) [inline app/src/main/java/com/kodeco/android/countryinfo/ui/components/CountryInfoScreen.kt]: If you remove the `Response` wrapper as I mentioned elsewhere you could simplify this code here too:

```suggestion
                infoState = try {
                    CountryInfoState.Success(service.getAllCountries()) {
                } catch (exception: Exception) {
                    CountryInfoState.Error(exception)
                }
```

**Review by zlippard — APPROVED** [review]

File: app/src/main/java/com/kodeco/android/countryinfo/api/CountryAPIService.kt (hunk 1 above)
Comment (savithaR24) [reply app/src/main/java/com/kodeco/android/countryinfo/api/CountryAPIService.kt]: Implemented this change and pushed. Helped reduce the number of the code lines and much simpler. Thank you.

File: app/src/main/java/com/kodeco/android/countryinfo/models/Country.kt (hunk 2 above)
Comment (savithaR24) [reply app/src/main/java/com/kodeco/android/countryinfo/models/Country.kt:15]: Thank you

File: app/src/main/java/com/kodeco/android/countryinfo/ui/components/CountryErrorScreen.kt (hunk 3 above)
Comment (savithaR24) [reply app/src/main/java/com/kodeco/android/countryinfo/ui/components/CountryErrorScreen.kt:31]: Yep, was aware of this issue. Tried doing the `connectivityManager` way but that proved complex for this app work. Hence, just added string check on the exception.

File: app/src/main/java/com/kodeco/android/countryinfo/ui/components/CountryInfoScreen.kt (hunk 4 above)
Comment (savithaR24) [reply app/src/main/java/com/kodeco/android/countryinfo/ui/components/CountryInfoScreen.kt]: Nice solution. Implemented and pushed. Thank you.

**Pull request merged**